File: backend/app/models/lifecycle_message.py
import uuid
from datetime import datetime

from sqlalchemy import String, Text, DateTime, ForeignKey, Integer
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base


class LifecycleMessage(Base):
    __tablename__ = "lifecycle_messages"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    chat_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("chats.id"))
    bot_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("bots.id"))
    session_id: Mapped[int] = mapped_column(Integer)  # Incremental ID for each proactive session
    role: Mapped[str] = mapped_column(String(20))  # "system", "assistant", or "tool"
    content: Mapped[str] = mapped_column(Text)
    content_type: Mapped[str] = mapped_column(String(50), default="text")  # text, tool_call, system_prompt
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)

    chat = relationship("Chat", foreign_keys=[chat_id])
    bot = relationship("Bot", foreign_keys=[bot_id])

File: app/lib/models/lifecycle_message.dart
class LifecycleMessage {
  final String id;
  final String chatId;
  final String botId;
  final int sessionId;
  final String role; // "system", "user", "assistant", or "tool"
  final String content;
  final String contentType; // "text", "tool_call", "tool_result", "system_prompt"
  final DateTime createdAt;

  const LifecycleMessage({
    required this.id,
    required this.chatId,
    required this.botId,
    required this.sessionId,
    required this.role,
    required this.content,
    required this.contentType,
    required this.createdAt,
  });

  factory LifecycleMessage.fromJson(Map<String, dynamic> json) {
    return LifecycleMessage(
      id: json['id'] as String,
      chatId: json['chat_id'] as String,
      botId: json['bot_id'] as String,
      sessionId: json['session_id'] as int,
      role: json['role'] as String,
      content: json['content'] as String,
      contentType: json['content_type'] as String,
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }
}
